### PR TITLE
WIP: Rework report – Looking for feedback!

### PR DESCRIPTION
**Pull request by hoechenberger**
## Motivation

The goal of this PR is to make `Report` more versatile and to update the UI and functionality to better meet the expectations of new users in 2021.

`Report` used to have only few public methods; the primary way to use it was via `parse_folder()`. This, however, gave users little control over the generated output. In the MNE-BIDS-Pipeline, we would render most objects manually and then add them via `add_figs_to_section()` – a somewhat dull and cumbersome endeavour.

I therefore decided it would be nice to have numerous public methods that have an inherent understanding of various MNE-Python objects and could render them in a sensible way automatically, without requiring users to cook up their own plotting code.

### Methods

- `add_epochs`
- `add_evokeds`
- `add_raw`
- `add_stc`
- `add_forward`
- `add_trans`
- `add_inverse`
- `add_covariance`
- `add_events`
- `add_ssp_projs`
- `add_code` (arbitrary code blocks with syntax highlighting)
- `add_sys_info` (output of `mne sys_info`)
- `add_bem` (replaces `add_bem_to_section`)
- `add_figure` (replaces `add_figs_to_section`)
- `add_image` (replaces `add_images_to_section`)
- `add_slider` (replaces `add_slider_to_section`)
- `add_html` (replaces `add_htmls_to_section`)

### "Data model"

The notion of "sections" is gone. Instead, all methods accept a parameter, `tags`, allowing users to assign an arbitrary number of tags to each added element.

### UI

- Table of contents now features a scrollspy, i.e. highlights the currently viewed section.
- Elements are collapsible by clicking on their title.
- New dropdown menu to select tags to display / hide.

### Dependencies, custom JavaScript & CSS

- I've updated jQuery and Bootstrap to the latest stable versions.
- Most of our custom CSS could be removed by styling via Bootstrap classes.
- I removed our custom slider, and created a new one based on Bootstrap's Carousel and Range components, with a little bit of custom JavaScript.
- I removed jquery-ui, which we needed only for our custom slider.
- I replaced jQuery calls with standard JS calls wherever possible.
- Custom JavaScript now makes use of some basic [ES6](https://www.w3schools.com/js/js_es6.asp) features, which have been supported by all major browsers for 5–10 years now (specifically, I'm using [fat arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#syntax), the `const` and `let` keywords, and the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax))

## MWE

```python
# %%
from pathlib import Path
import mne


root_dir = Path(mne.datasets.sample.data_path())
meg_dir = root_dir / 'MEG' / 'sample'
fs_subjects_dir = root_dir / 'subjects'
fs_subject = 'sample'

raw_path = meg_dir / 'sample_audvis_filt-0-40_raw.fif'
ecg_proj_path = meg_dir / 'sample_audvis_ecg-proj.fif'
eog_proj_path = meg_dir / 'sample_audvis_eog-proj.fif'
trans_path = meg_dir / 'sample_audvis_raw-trans.fif'
cov_path = meg_dir / 'sample_audvis-cov.fif'
fwd_path = meg_dir / 'sample_audvis-meg-oct-6-fwd.fif'
inverse_path = meg_dir / 'sample_audvis-meg-oct-6-meg-inv.fif'
stc_path = meg_dir / 'sample_audvis-meg'

# Create epochs
raw = mne.io.read_raw(raw_path)
events = mne.find_events(raw, stim_channel='STI 014')
event_name_id_map = {
    'auditory/left': 1, 'auditory/right': 2, 'visual/left': 3,
    'visual/right': 4, 'smiley': 5, 'buttonpress': 32
}
epochs = mne.Epochs(
    raw=raw, events=events, event_id=event_name_id_map,
    tmin=-0.2, tmax=0.5, preload=True
)

evokeds = [
    epochs['auditory'].average(),
    epochs['visual'].average()
]

# %%
r = mne.Report(title='My Report')

r.add_raw(raw=raw_path, title='my raw data', tags=('raw',))
r.add_events(events=events, event_id=event_name_id_map, title='my events',
             sfreq=raw.info['sfreq'])
r.add_epochs(epochs=epochs, title='my epochs', tags=('epochs',))
r.add_evokeds(evokeds=evokeds[:2], noise_cov=cov_path,
              titles=['my evoked 1', 'my evoked 2'],
              tags=('evoked',))
r.add_ssp_projs(info=evokeds[0].info, projs=ecg_proj_path,
                title='my proj', tags=('ssp', 'ecg'))
r.add_ssp_projs(info=evokeds[0].info, projs=eog_proj_path,
                title='my proj', tags=('ssp', 'eog'))
r.add_covariance(cov=cov_path, info=raw_path, title='my cov')
r.add_trans(trans=trans_path, info=raw_path, title='my coreg',
            subject=fs_subject, subjects_dir=fs_subjects_dir)
r.add_bem(subject=fs_subject, subjects_dir=fs_subjects_dir,
          title='my bem')
r.add_forward(forward=fwd_path, title='my forward',
              subject=fs_subject, subjects_dir=fs_subjects_dir)
r.add_inverse(inverse=inverse_path, title='my inverse',
              subject=fs_subject, subjects_dir=fs_subjects_dir,
              trans=trans_path)
r.add_stc(stc=stc_path, title='my stc',
          subject=fs_subject, subjects_dir=fs_subjects_dir)
r.add_code(code=__file__, title='my code')
r.add_sys_info(title='my sysinfo')

r.save('/tmp/report.html', open_browser=True, overwrite=True)
```

Screen Shot 2021-09-07 at 21 03 59

## Issues & todo

- [x] Ensure existing tests pass
- [x] Add tests for new methods (63681d9)
- [x] The tags dropdown needs a scrollbar if there are many tags (a715cf2)
- [ ] Update tutorial
- [x] Automatic ordering after `parse_folder()` run
- [x] Saving to and loading from HDF5 is broken (3170593)
- [x] Methods that need to be updated to match the new data model (and also should be deprecated):
  - [x] `add_figs_to_section` (250ebda)
  - [x] `add_images_to_section` (ad496a7)
  - [x] `add_slider_to_section` (5ff87cc)
  - [x] `add_htmls_to_section` (7ffea91)
  - [x] `add_bem_to_section` (987706b)
- [x] Scrollspy sometimes doesn't work properly in certain situations (c721d26)
- [x] <strike>Currently, each condition in an evoked file gets its own TOC entry. In `main`, there is some kind of nesting. Do we want / need this?</strike>
- [x] I pass around `data_path` in many places where we don't actually need it anymore (a8fa634)
- [x] Ensure `replace` parameter exists & works (6e534d7)
- [x] Fix `Report.remove()` (deb88ea)
- [x] <strike>Better input sanitization</strike>
- [x] <strike>Replace unmaintained Tempita templating engine with Jinja</strike >

### Low prio / do later

- [ ] Clicking on a tag should hide all other content
- [ ] Reconsider layout -> make columns truly responsive
- [ ] On small screens, the TOC on the left should be hidden in a hamburger menu

The code, all in all, while functional, is still kind of messy. Still I'd appreciate code design feedback if you dare :) Otherwise I'm super happy about any feedback and thoughts on the redesigned UI and UX.

cc @dengemann @SophieHerbst @drammock @agramfort @sappelhoff @jasmainak

### CODE REVIEW

File: mne/viz/utils.py [hunk 1]
```diff
@@ -1609,7 +1609,7 @@ def _setup_ax_spines(axes, vlines, xmin, xmax, ymin, ymax, invert_y=False,
     # new ticks that are nice round numbers close to (but less extreme than)
     # xmin and xmax
     vlines = [] if vlines is None else vlines
-    xticks = _trim_ticks(axes.get_xticks(), xmin, xmax)
+    xticks = _trim_ticks(axes.get_xticks(), round(xmin, 2), round(xmax, 2))
```
Comment (hoechenberger) [inline mne/viz/utils.py:1612]: Whoops, this should probably go to a separate PR. It improves the X-axis visualization of the GFP plots.

**Comment by hoechenberger** [comment]
- We're also adding `alt` tags in many more places and I'm trying to follow a11y guidelines as good as I can

**Comment by rob-luke** [comment]
Reports are awesome and I'm currently addicted to them. The screen shot above looks great, and I like the tags filter feature.

It's unclear to me if you are proposing to depreciate `add_figs_to_section` or just revamp it? If you are depreciating, then how would one achieve the same functionality with your new methods?

**Comment by jasmainak** [comment]
Fantastic initiative! I second @rob-luke that I kind of like `add_figs_to_section`. For example, I could iterate over all the open figures in matplotlib and create a report that shows me all the figures from a piece of script.

I almost never need any of the other functions *except* for `trans` (and BEM). You need three views to get a good idea of the co-registration and messing with mayavi code to do that is somewhat non-trivial.

**Comment by jasmainak** [comment]
The tags feature looks awesome :-)

**Comment by hoechenberger** [comment]
Hello @jasmainak and @rob-luke, thanks for the positive feedback!

My goal is to retain backward-compatibility as much as possible, so my immediate next steps are 

- to ensure that all "old" `add_*_to_section()` methods still work, and simply map "section" to "tags"
- add new methods that implement the same functionality, but follow the new method naming / signatures
- deprecate the old methods, possibly with a longer-than-usual deprecation cycle if you think that makes sense

Does that sound okay to you?

**Comment by rob-luke** [comment]
Sounds great. Let me know when you want me to give it a spin locally.

**Review by sappelhoff — COMMENTED** [review]
I haven't had the chance to use reports as much as I would have liked to, but the proposed changes sound good to me - and I'll try to give this a spin in my next project (couple of months/weeks from now though).

**Comment by larsoner** [comment]
Looks very promising!

We should definitely keep some variant of `add_figs_to_section`, being able to add custom items not supported directly in MNE is very useful. At UW we use a lot of custom report-related code -- once you think you have this backward compatibility worked out, I can try running our report generation code and fix any bugs I hit along the way (or report them here if you want to do it, but it's probably easier for me since I have the custom-report-generation code all set up locally).

Currently the diff is a pain to look at because it shows up as `mne/report.py` being deleted and `mne/report/report.py` being created. Could you at least for now restructure it so that there is `mne/report.py` with all the report code in it, and then have `mne/_report/*` for all the other stuff? That way reviewing the Python code changes will be much easier. Then right before merge merge, hopefully it's quick and easy to go back to the structure you have here just by changing a few paths (which I agree is better than a single monolithic `report.py`). If it's not clear what I mean, I can push a quick commit to do this.

**Comment by hoechenberger** [comment]
> Currently the diff is a pain to look at because it shows up as `mne/report.py` being deleted and `mne/report/report.py` being created.

Yes, I'm surprised it is that way though, I thought `git` would track the content :( But even without this, the diff would be super painful, as I touch(ed) almost everything in `report.py` :S But I will see if I can make things slightly better by following your proposal.

I've locally gotten `add_figs_to_section()` to work again; it internally uses a new `add_figs()`. Still fighting with a last remaining bug.

While we're talking about this:
Are the `scale`, `replace`, and `auto_close` params actually of any use to you?

**Comment by larsoner** [comment]
> Are the scale, replace, and auto_close params actually of any use to you?

I didn't even know those existed :) And based on what I assume they do, no, they are not. Scale and close can be done by the user when creating their figures. Replace maybe is useful but when scripting these things usually you don't need it.

Really @wmvanvliet might have some opinion on these, though, since I know he worked on class persistence maybe `replace` is indeed useful in his workflow(s)

**Comment by hoechenberger** [comment]
I've now resurrected `add_figs_to_section()`, which internally calls the new method `add_figs()`.

MWE:
```python
# %%
import mne
import matplotlib.pyplot as plt

r = mne.Report()

fig, ax = plt.subplots()
ax.plot([1,2,3], [1,2,3])

r.add_figs_to_section(
    figs=fig,
    captions='Old API, single fig',
    section='old-api',
    comments='this is a comment'
)

r.add_figs_to_section(
    figs=[fig, fig],
    captions=['Old API, multiple figs (1)',
              'Old API, multiple figs (2)'],
    section='old-api',
    comments=['a comment', 'this is another comment']
)


# New API, multiple figs
r.add_figs(figs=[fig, fig],
           titles=['New API (1)', 'New API (2)'],
           captions=['A caption', 'Another caption'],
           tags=('new-api',)
)


r.save('/tmp/report_figs.html', overwrite=True)
```

Rendered HTML can be downloaded from my Google Drive:
https://drive.google.com/file/d/1gmyZWApXOi9BhJovPDvEQVx3Sl26MxnA

@jasmainak @larsoner @rob-luke Does this have the functionality you need? Or did I mess something up for you?


Next up: fix the remaining `add_*_to_section()` methods!

**Comment by larsoner** [comment]
> Rendered HTML can be downloaded from my Google Drive:

I would rather see the CircleCI build fixed, then we can review using the standard CI checking schemes. `add_figs_to_section` is used in `tutorials/intro/70_report.py`...

**Comment by hoechenberger** [comment]
> I would rather see the CircleCI build fixed

Good point, coming right up! 🏃 🏃 🏃

**Comment by drammock** [comment]
one minor point of feedback: I noticed (in your google-drive-shared example of the old and new APIs) that when you deselect "old API" tag, the old-api-tagged elements disappear from the main body of the report (as expected) but their titles also disappear from the left-side TOC (not expected... I think I would want them to be disabled / greyed out maybe, rather than completely hidden?).  This opinion is not super-strong, however, so I'm prepared to be talked out of it.

**Comment by hoechenberger** [comment]
@drammock Totally open for discussion on this one! We'd need to see how the scrollspy behaves if we do that the way you describe, though. But I would think that it's doable…

**Comment by larsoner** [comment]
`[circle full]` will take 2h to run, it's better to touch the 1 or 2 examples that you actually want to run and let CircleCI's automatic "which examples should I execute based on the diff" code do its magic

**Comment by hoechenberger** [comment]
@larsoner I actually felt that it was a good idea to force a full run; but now I'm not convinced anymore 🙈 Will touch the relevant tutorials to avoid such lengthy runs in the future. Sorry about that!

**Comment by wmvanvliet** [comment]
> Are the `scale`, `replace`, and `auto_close` params actually of any use to you?

`replace` is crucial to my workflow, because I add to the same report from multiple scripts. Re-running a script should replace some figures in the bigger report.

The new UI looks great!

**Comment by hoechenberger** [comment]
Thanks, @wmvanvliet! I will ensure that `replace` works as expected (currently not the case)

**Comment by hoechenberger** [comment]
@larsoner Circle still doesn't finish fully, but the Report tutorial does get rendered:
https://33866-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/intro/70_report.html

One can see that I still need to do some adjustments to the layout so it will look good on small screens

**Comment by hoechenberger** [comment]
Alright, doc build is working for now!

**Comment by hoechenberger** [comment]
All "old" methods are working again now, save for the `replace` parameter (which is currently silently ignored). This is the next big thing on my todo list, along with HDF5 im- and export.

File: mne/viz/tests/test_topo.py [hunk 2]
```diff
@@ -89,7 +89,7 @@ def return_inds(d):  # to test function kwarg to zorder arg of evoked.plot
     axes = plt.subplots(nrows=3)[-1].flatten().tolist()
     evoked.plot_joint(times=[0], picks=[6, 7, 8], ts_args=dict(axes=axes[0]),
                       topomap_args={"axes": axes[1:], "time_unit": "s"})
-    with pytest.raises(ValueError, match='array of length 6'):
+    with pytest.raises(ValueError, match='array of length 4'):
```
Comment (hoechenberger) [inline mne/viz/tests/test_topo.py:92]: This was actually a bug in the test

**Comment by hoechenberger** [comment]
@drammock I'm having a bit of trouble with my reworked `viz.misc._plot_mri_contours()` function. I managed to narrow down the problem, and it's related to an interaction between Matplotlib 3.1 and joblib. Here is a MWE:


```python
# %%
from mne.parallel import parallel_func
import matplotlib
import matplotlib.pyplot as plt

matplotlib.use('Agg')


def create_fig():
    fig, ax = plt.subplots()
    plt.close(fig)
    return fig


parallel, p_fun, _ = parallel_func(
    func=create_fig,
    n_jobs=2
)

figs = parallel(
    p_fun() for _ in range(10)
)

print([f.canvas for f in figs])
```

This prints:
```
[None, None, None, None, None, None, None, None, None, None]
```

If I change `n_jobs` to `1`, I get the expected output:
```python
[<matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6e740650>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6e8fbcd0>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6e93dc10>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6e97cb90>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6e9bcb10>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6eb21a90>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6eb62a10>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6eb9e950>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6ebde890>, <matplotlib.backends.backend_agg.FigureCanvasAgg object at 0x7f8d6ed1d810>]
```

I get similar (desired) output for `n_jobs > 1` with Matplotlib >3.1.

I wanted to ask if you have any idea what could be going on? I couldn't find anything to this particular behavior change in their changelog :(

**Comment by hoechenberger** [comment]
@drammock never mind the above comment, I think I've found a solution that I will try out momentarily.

File: mne/report/tests/test_report.py [hunk 3]
```diff
@@ -400,7 +400,7 @@ def test_add_bem_n_jobs(n_jobs, monkeypatch):
     norms = np.linalg.norm(imgs, axis=-1)
     # should have down-up-down shape
     corr = np.corrcoef(norms, np.hanning(len(imgs)))[0, 1]
-    assert 0.78 < corr < 0.80
+    assert 0.778 < corr < 0.80
```
Comment (hoechenberger) [inline mne/report/tests/test_report.py:403]: For passing locally on my Mac

File: mne/viz/misc.py [hunk 4]
```diff
@@ -302,9 +302,22 @@ def plot_source_spectrogram(stcs, freq_bins, tmin=None, tmax=None,
...
-                       slices=None, show=True, show_indices=False,
-                       show_orientation=False, img_output=False, width=512):
-    """Plot BEM contours on anatomical slices."""
+                       slices=None, slices_as_figures=False, show=True,
+                       show_indices=False, show_orientation=False, width=512):
+    """Plot BEM contours on anatomical slices.
+
+    Parameters
+    ----------
+    slices_as_figures : bool
```
Comment (hoechenberger) [inline mne/viz/misc.py:311]: I document the new param I added, but didn't feel the urge to add docs for all the other params there were already undocumented before I came along 😅

File: mne/report/js_and_css/report.sass [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+#content {
+    margin-top: 90px;
+    scroll-behavior: smooth;
+    position: relative;  // for scrollspy
+}
+
+#toc {
+    margin-top: 90px;
+    padding-bottom: 5rem;
```
Comment (drammock) [inline mne/report/js_and_css/report.sass:14]: sass does not use semicolons

File: mne/report/report.py [hunk 6]
```diff
@@ -0,0 +1,3203 @@
...
+
+def _get_mri_contour_figs(*, sl, n_jobs, mri_fname, surfaces,
+                          orientation, src, show, show_orientation, width,
+                          slices_as_figures):
+    import matplotlib.pyplot as plt
+
+    # Matplotlib <3.2 doesn't work nicely with process-based parallelization
+    from matplotlib import __version__ as MPL_VERSION
+    from pkg_resources import parse_version
+    if parse_version(MPL_VERSION) >= parse_version('3.2'):
```
Comment (drammock) [inline mne/report/report.py:256]: elsewhere we mostly use `from distutils.version import LooseVersion` to do comparisons like this.  Is this approach better / more modern?

Comment (larsoner) [reply mne/report/report.py:256]: `distutils` is [being deprecated](https://www.python.org/dev/peps/pep-0632). +1 for adding a `mne.fixes._version_greater(a, b, comp='>=')` or something that uses `pkg_resources` if available and `LooseVersion` if not (`comp` could be `'>='` or `'>'`)

Comment (hoechenberger) [reply mne/report/report.py:256]: Great idea @larsoner, will give it a shot. The deprecation was actually the reason why I picked this approach, yes.

Comment (hoechenberger) [reply mne/report/report.py]: @larsoner Please take a look at f29eced and let me know what you think! :)

File: mne/fixes.py [hunk 7]
```diff
@@ -1077,3 +1076,36 @@ def _cell_data(obj):
...
+
+###############################################################################
+# distutils
+
+# distutils has been deprecated since Python 3.10 and is scheduled for removal
+# from the standard library with the release of Python 3.12. For version
+# comparisons, we use setuptools's `parse_version` if available.
+
+def _compare_version(version_a, operator, version_b):
+    """Check if
```
Comment (hoechenberger) [inline mne/fixes.py:1089]:
```suggestion
    """Compare two version strings via a user-specified operator.
```

File: mne/fixes.py [hunk 8]
```diff
@@ -1077,3 +1076,36 @@ def _cell_data(obj):
...
+# comparisons, we use setuptools's `parse_version` if available.
+
+def _compare_version(version_a, operator, version_b):
+    """Check if 
+
+    Parameters
+    ----------
+    version_a : str
+        First version string.
+    operator : '==' | '>' | '<' | '>=' | '<=' | '~='
```
Comment (hoechenberger) [inline mne/fixes.py:1095]:
```suggestion
    operator : '==' | '>' | '<' | '>=' | '<='
```

File: mne/report/js_and_css/report.sass (hunk 5 above)
Comment (hoechenberger) [reply mne/report/js_and_css/report.sass]: Thank you, addressed in 3b1a73e

File: mne/report/report.py (hunk 6 above)
Comment (hoechenberger) [reply mne/report/report.py]: @larsoner FYI I made a followup commit to add some fixes, b6f4f5c

File: mne/fixes.py [hunk 9]
```diff
@@ -23,6 +22,39 @@
...
 
+###############################################################################
+# distutils
+
+# distutils has been deprecated since Python 3.10 and is scheduled for removal
+# from the standard library with the release of Python 3.12. For version
+# comparisons, we use setuptools's `parse_version` if available.
+
+def _compare_version(version_a, operator, version_b):
+    """Compare two version strings via a user-specified operator;
```
Comment (hoechenberger) [inline mne/fixes.py:33]:
```suggestion
    """Compare two version strings via a user-specified operator.
```

**Comment by larsoner** [comment]
@hoechenberger let me know when I should look

**Comment by hoechenberger** [comment]
The only thing left on my todo list (top post) is "Update tutorial".

@larsoner I've tried to rework the git history a little by soft-resetting the branch to cb1916f0e7556d6b84e300f0e26bfceabd972110 (the last commit before I moved `report.py` to a new sub-directory) and committing all the changes done thereafter, hoping that this would preserve the history of the file's content. While this seems to be doing "the right thing" locally, GH doesn't appear to agree and still insists that `report.py` was deleted and a new file, `report/report.py`, was created from scratch.

If you have any advice / would like to give this a shot, please feel free to play with and force-push to this branch.

**Comment by larsoner** [comment]
Closing in favor of your next upcoming PR
